File: lib/repository/comment/json.py
from datetime import datetime;
import uuid;
import json;

# Implements IRepository interface for reading/writing comments to a JSON file.
# See interfaces/comment_repository for method documentation


class CommentJSONRepository():

  def __init__(self, file_path="data.json"):
    self.__file_path = file_path;
    with open(file_path, "r") as json_file:
      self.__db = json.load(json_file);


  def create(self, doc):
    my_uuid = str(uuid.uuid4());

    self.__db[my_uuid] = {
      "id": my_uuid,
      "createdDate": str(datetime.now()),
      "lastModified": "n/a",
      "doc": doc
    };

    self.save_to_json_file();
    return my_uuid;


  def find_one(self, id):
    return self.__db[id];


  def find_all(self):
    return self.__db.values();


  def update_comment(self, id, doc):
    #self.__db[id]["doc"].update(doc);
    #self.__db[id].update({"lastModified": str(datetime.now())});
    #self.save_to_json_file();
    pass;


  def save_to_json_file(self):
    doc = json.dumps(self.__db, indent=2);
    with open(self.__file_path, "w") as json_file:
      json_file.write(doc);



####JSONRepository####

File: src/lib/repository/comment/json.js
/* istanbul ignore file */
const {CommentDTO, CommentLikeDTO} = require("../../repository/comment/dto");

/*Implements ICommentRepository interface for connecting to JSON file of in memory document database
*See interfaces/comment-repository for method documentation
*/

/**
 * @implements {ICommentRepository}
 */

function CommentJSONRepository(databaseConnector) {

    /**
     * 
     * @param {CommentDTO} commentDTO - an instance of CommentDTO 
     * @param {CommentLikeDTO} commentLikeDTO - an instance of CommentLikeDTO 
     */
    this.create = async function(commentDTO, commentLikeDTO) {
        const [comment] = await databaseConnector.add({
            doc: commentDTO, 
            collection: "comments"
        });

        await databaseConnector.putOne({
            doc: commentLikeDTO, 
            collection: "comment_likes"
        });

        return { id: comment.id, createdDate: comment.createdDate };
    }


    this.findOne = async function(id) {
        const result = await databaseConnector.findOne({id, collection:"comments"});
        return result;
    }


    this.findAllComments = async function() {
        const result = await databaseConnector.findAll("comments");
        return result;
    }

    /**
     * 
     * @param {CommentDTO} commentDTO - an instance of CommentDTO 
     * @param {String} text - text to update comment
     */

    this.editComment = async function(commentDTO, text) {
        const commentData = commentDTO.value();

        const [record] = await databaseConnector.updateOne({
            doc: new CommentDTO(Object.assign(commentData, {
                body: text
            })), 
            collection: "comments"
        });

        return { id: record.id , lastModified: record.lastModified };
    }

    /**
     * 
     * @param {CommentDTO} commentDTO - an instance of CommentDTO 
     * @param {String} fromUser - uuid of platform user liking the comment
     */
    this.incrementLikeCount = async function(commentDTO, fromUser) {
        const commentData = commentDTO.value();
        const commentId = commentData.id;
        const [commentLike] = await databaseConnector.findOne({
            id: commentId, 
            collection: "comment_likes"
        });
        const commentLikeListUnique = new Set(commentLike.likes);
        commentLikeListUnique.add(fromUser);
        
        const updatedCommentDTO = new CommentDTO(Object.assign(commentData, {
            likeCount: commentLikeListUnique.size
        }));
        const updatedCommentLikeDTO = new CommentLikeDTO(Object.assign(commentLike, {
            likes: Array.from(commentLikeListUnique)
        }))
        
        await databaseConnector.updateOne({
            doc: updatedCommentDTO,
            collection: "comments"
        });

        await databaseConnector.updateOne({
            doc: updatedCommentLikeDTO,
            collection: "comment_likes"
        });
    }

    this.decrementLikeCount = async function(commentDTO, fromUser) {
        const commentData = commentDTO.value();
        const commentId = commentData.id;
        const [commentLike] = await databaseConnector.findOne({
            id: commentId, 
            collection: "comment_likes"
        });
        const commentLikeListUnique = new Set(commentLike.likes);
        
        commentLikeListUnique.delete(fromUser);

        const updatedCommentDTO = new CommentDTO(Object.assign(commentData, {
            likeCount: commentLikeListUnique.size
        }));
        const updatedCommentLikeDTO = new CommentLikeDTO(Object.assign(commentLike, {
            likes: Array.from(commentLikeListUnique)
        }))

        await databaseConnector.updateOne({
            doc: updatedCommentDTO,
            collection: "comments"
        });

        await databaseConnector.updateOne({
            doc: updatedCommentLikeDTO,
            collection: "comment_likes"
        });
    }

}


/*CommentJSONRepository*/

module.exports = CommentJSONRepository;
